r1cs/gadgets/crypto/src/crh/poseidon/tweedle/density_optimized/mod.rs

```diff
@@ -0,0 +1,525 @@
+use crate::FieldBasedHashGadget;
+use algebra::fields::{tweedle::Fr, Field};
+use primitives::crh::{FieldBasedHashParameters, PoseidonParameters, TweedleFrPoseidonHash};
+use r1cs_core::{ConstraintSystem, ConstraintVar, LinearCombination, SynthesisError};
+use r1cs_std::{
+    alloc::{AllocGadget, ConstantGadget},
+    fields::{fp::FpGadget, FieldGadget},
+    Assignment,
+};
+
+mod constants;
+use constants::*;
+
+/// Poseidon Hash implementation optimized for densities, in this
+/// very particular use case (e.g. SBox = x^5, R_F = 8 and R_P = 56).
+/// Might be generalizable, for the moment it isn't and it's not of
+/// our interest to do so, but we tried to reduce as much as possible
+/// the usage of magic numbers in favor of trait constants and use
+/// generic functions.
+pub struct TweedleFrDensityOptimizedPoseidonHashGadget {}
```

Comment:
As pointed out offline, I'd opt for another name here. Every circuit in our repository will be for Marlin, beside few exceptions. So let us choose shorter names in standard use cases, and longer names for the exceptions.

Reply:
Let's address this in another PR in which we will generalize this design.